core/domain/stats_services_test.py

```diff
@@ -14,20 +14,233 @@
+        handle_stats_creation_for_new_exploration methods when an exploration is
+        created or updated.
+        """
+        # Create exploration object in datastore.
+        exp_id = 'exp_id'
+        test_exp_filepath = os.path.join(
+            feconf.TESTS_DATA_DIR, 'string_classifier_test.yaml')
+        yaml_content = utils.get_file_contents(test_exp_filepath)
+        assets_list = []
+        with self.swap(feconf, 'ENABLE_ML_CLASSIFIERS', True):
+            exp_services.save_new_exploration_from_yaml_and_assets(
+                feconf.SYSTEM_COMMITTER_ID, yaml_content, exp_id,
+                assets_list)
+        exploration = exp_services.get_exploration_by_id(exp_id)
+
+        # Now, there are two analytics model in datastore (one from setup
+        # function).
+        all_models = stats_models.ExplorationStatsModel.get_all()
+        self.assertEqual(all_models.count(), 2)
+        exploration_stats = stats_services.get_exploration_stats_by_id(
+            exploration.id, exploration.version)
+        self.assertEqual(
+            exploration_stats.state_stats_mapping.keys(), ['Home', 'End'])
+
+        # Update exploration by adding a state.
+        change_list = [{
+            'cmd': 'add_state',
+            'state_name': 'New state'
+        }]
+        exploration.add_states(['New state'])
+        exploration.version += 1
+        exp_services.update_exploration(
+            feconf.SYSTEM_COMMITTER_ID, exp_id, change_list, '')
+
+        # Now, there should be three analytics models in datastore (one from
+        # setup function).
+        all_models = stats_models.ExplorationStatsModel.get_all()
+        self.assertEqual(all_models.count(), 3)
+        exploration_stats = stats_services.get_exploration_stats_by_id(
+            exploration.id, exploration.version)
+        self.assertEqual(
+            exploration_stats.state_stats_mapping.keys(), [
+                'Home', 'End', 'New state'])
+
+    def test_handle_stats_creation_for_new_exploration(self):
+        """Test the handle_stats_creation_for_new_exploration method."""
+        # Create exploration object in datastore.
+        exp_id = 'exp_id'
+        test_exp_filepath = os.path.join(
+            feconf.TESTS_DATA_DIR, 'string_classifier_test.yaml')
+        yaml_content = utils.get_file_contents(test_exp_filepath)
+        assets_list = []
+        with self.swap(feconf, 'ENABLE_ML_CLASSIFIERS', True):
+            exp_services.save_new_exploration_from_yaml_and_assets(
+                feconf.SYSTEM_COMMITTER_ID, yaml_content, exp_id,
+                assets_list)
+        exploration = exp_services.get_exploration_by_id(exp_id)
+
+        stats_services.handle_stats_creation_for_new_exploration(exploration)
+
+        exploration_stats = stats_services.get_exploration_stats_by_id(
+            exploration.id, exploration.version)
+        self.assertEqual(
+            exploration_stats.state_stats_mapping.keys(), ['Home', 'End'])
```

Comment:
What about testing the rest of the object?

Reply:
Done.